### ISSUE TITLE
False-positive E0602 with nonlocal

### ISSUE THREAD

**Issue by oriyarde** [PyCQA/pylint#5785]
### Bug description

similar to https://github.com/PyCQA/pylint/issues/2049, but with `with` instead of `if`:

```python
import threading


def func():
    count = 0
    lock = threading.Lock()
    def inner():
        with lock:
            nonlocal count
            count += 1
    inner()
```

### Command used

```shell
pylint nonlocalint.py
```

### Pylint output

```shell
************* Module nonlocalint
nonlocalint.py:10:12: E0602: Undefined variable 'count' (undefined-variable)
```

### Expected behavior

No error. `count` equals 1 at the end of `func`

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.3
Python 3.8.8 (default, Aug 11 2021, 06:52:42)
```

**Comment by mbyrnepr2** [comment]
I haven't spotted the change which fixed this but I can reproduce on the reported 2.12.2 version but not on development 2.13.0

**Comment by DanielNoord** [comment]
Thanks @mbyrnepr2! We can probably thank @jacobtylerwalls for this!

Closing!
